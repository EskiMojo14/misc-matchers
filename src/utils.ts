import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}

/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
