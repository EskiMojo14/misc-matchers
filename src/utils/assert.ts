export function assert(
  condition: unknown,
  message: string | (() => string),
): asserts condition {
  if (!condition) {
    throw new Error(typeof message === "function" ? message() : message);
  }
}
