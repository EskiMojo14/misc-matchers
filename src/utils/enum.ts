export interface EnumLike {
  [key: string]: string | number;
  [index: number]: string;
}

export const getValidEnumValues = (obj: EnumLike) =>
  Object.values(obj).filter((v) => typeof obj[v] !== "number");
