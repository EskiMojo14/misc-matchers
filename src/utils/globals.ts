// to avoid having separate jest and vitest suites, we create common definitions
// then mock during setup with the appropriate globals
import type {
  describe as jestDescribe,
  it as jestIt,
  expect as jestExpect,
  jest,
} from "@jest/globals";
import type {
  describe as vitestDescribe,
  it as vitestIt,
  expect as vitestExpect,
  vi,
  Mock as VitestMock,
} from "vitest";
import type { Satisfies } from "./types.typetest";

interface Each {
  <T extends Array<any> | [any]>(
    cases: ReadonlyArray<T>,
  ): (name: string, fn: (...args: T) => void) => void;
  <T>(cases: ReadonlyArray<T>): (name: string, fn: (value: T) => void) => void;
}
export declare const describe: (typeof vitestDescribe | typeof jestDescribe) & {
  each: Each;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestEach = Satisfies<
  typeof jestDescribe.each | typeof vitestDescribe.each,
  Each
>;

export declare const expect: typeof vitestExpect | typeof jestExpect;
export declare const it: typeof vitestIt | typeof jestIt;

export type Mock<R = any, A extends Array<any> = Array<any>, T = any> =
  | jest.Mock<(this: T, ...args: A) => R>
  | VitestMock<A, R>;

export declare const fn: {
  (): Mock;
  <R, A extends Array<any>, T>(fn?: (this: T, ...args: A) => R): Mock<R, A, T>;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type TestFn = Satisfies<typeof jest.fn | typeof vi.fn, typeof fn>;

throw new Error("This file should be mocked");
