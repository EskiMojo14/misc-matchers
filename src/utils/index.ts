import type { MatcherHintOptions } from "jest-matcher-utils";
import type { MatcherUtils } from "expect";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isMock = (received: any): received is jest.Mock =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  received?._isMockFunction === true;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isSpy = (received: any): received is jasmine.Spy =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.all === "function" &&
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  typeof received?.calls?.count === "function";

export function ensureMockOrSpy(
  utils: MatcherUtils["utils"],
  received: unknown,
  matcherName: string,
  expectedArgument?: string,
  options?: MatcherHintOptions,
): asserts received is jest.Mock | jasmine.Spy {
  if (!isMock(received) && !isSpy(received)) {
    throw new Error(
      utils.matcherErrorMessage(
        utils.matcherHint(matcherName, undefined, expectedArgument, options),
        `${utils.RECEIVED_COLOR(
          "received",
        )} value must be a mock or spy function`,
        utils.printWithType("Received", received, utils.printReceived),
      ),
    );
  }
}
