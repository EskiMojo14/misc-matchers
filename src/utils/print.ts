/**
 * Given a label, return a function which given a string,
 * right-aligns it preceding the colon in the label.
 */
export const getRightAlignedPrinter = (label: string) => {
  // Assume that the label contains a colon.
  const index = label.indexOf(":");
  const suffix = label.slice(index);

  return (string: string, isExpectedCall: boolean) =>
    (isExpectedCall
      ? `->${" ".repeat(Math.max(0, index - 2 - string.length))}`
      : " ".repeat(Math.max(index - string.length))) +
    string +
    suffix;
};
