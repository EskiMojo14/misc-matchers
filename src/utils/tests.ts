/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import ansiRegex from "ansi-regex";
// ansi-styles is ESM - as a result we need to run our Jest tests with `--experimental-vm-modules` flag
import style from "ansi-styles";
import type { NewPlugin } from "pretty-format";

export const alignedAnsiStyleSerializer: NewPlugin = {
  serialize(val: string | Error): string {
    // Return the string itself, not escaped nor enclosed in double quote marks.
    return (val instanceof Error ? val.message : val).replace(
      ansiRegex(),
      (match) => {
        switch (match) {
          case style.inverse.open:
            return "<i>";
          case style.inverse.close:
            return "</i>";

          case style.bold.open:
            return "<b>";
          case style.dim.open:
            return "<d>";
          case style.green.open:
            return "<g>";
          case style.red.open:
            return "<r>";
          case style.yellow.open:
            return "<y>";
          case style.bgYellow.open:
            return "<Y>";

          case style.bold.close:
            return "</b>";
          case style.dim.close:
            return "</d>";
          case style.green.close:
            return "</g>";
          case style.red.close:
            return "</r>";
          case style.yellow.close:
            return "</y>";
          case style.bgYellow.close:
            return "</Y>";

          default:
            return match; // unexpected escape sequence
        }
      },
    );
  },
  test(val: unknown) {
    return typeof val === "string" || val instanceof Error;
  },
};
