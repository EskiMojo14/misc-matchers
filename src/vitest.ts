/* eslint-disable @typescript-eslint/no-empty-interface */
import { expect } from "vitest";
import * as mixNMatchers from ".";

declare module "vitest" {
  interface Assertion<T> extends mixNMatchers.MixNMatchers<void, T> {}
  interface AsymmetricMatchersContaining
    extends mixNMatchers.AsymmetricMixNMatchers {
    enum: mixNMatchers.AsymmetricMixNMatchers["ofEnum"];
  }
}

expect.extend(mixNMatchers);
expect.extend({ enum: mixNMatchers.ofEnum });
