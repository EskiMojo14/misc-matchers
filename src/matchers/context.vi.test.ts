/* eslint-disable @typescript-eslint/no-empty-function */
import { expect, describe, it, vi } from "vitest";
import { alignedAnsiStyleSerializer } from "../utils/tests";

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

describe.each([
  "toHaveBeenCalledWithContext",
  "toBeCalledWithContext",
  "lastCalledWithContext",
  "toHaveBeenLastCalledWithContext",
  "nthCalledWithContext",
  "toHaveBeenNthCalledWithContext",
] as const)("%s", (calledWithContext) => {
  function isToHaveNth(
    calledWithContext: string,
  ): calledWithContext is
    | "nthCalledWithContext"
    | "toHaveBeenNthCalledWithContext" {
    return (
      calledWithContext === "nthCalledWithContext" ||
      calledWithContext === "toHaveBeenNthCalledWithContext"
    );
  }
  it("works only on spies or vi.fn", () => {
    const fn = function fn() {};
    if (isToHaveNth(calledWithContext)) {
      expect(() => {
        expect(fn)[calledWithContext](3, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(() => {
        expect(fn)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works when not called", () => {
    const fn = vi.fn();
    if (isToHaveNth(calledWithContext)) {
      expect(fn).not[calledWithContext](1, "foo");

      expect(() => {
        expect(fn)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn).not[calledWithContext]("foo");

      expect(() => {
        expect(fn)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match", () => {
    const fn = vi.fn();
    fn.call("bar");

    if (isToHaveNth(calledWithContext)) {
      expect(fn).not[calledWithContext](1, "foo");

      expect(() => {
        expect(fn)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn).not[calledWithContext]("foo");

      expect(() => {
        expect(fn)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match with matchers", () => {
    const fn = vi.fn();
    fn.call("bar");

    if (isToHaveNth(calledWithContext)) {
      expect(fn).not[calledWithContext](1, expect.any(Number));

      expect(() => {
        expect(fn)[calledWithContext](1, expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn).not[calledWithContext](expect.any(Number));

      expect(() => {
        expect(fn)[calledWithContext](expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches", () => {
    const fn = vi.fn();
    fn.call("foo");

    if (isToHaveNth(calledWithContext)) {
      expect(fn)[calledWithContext](1, "foo");

      expect(() => {
        expect(fn).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn)[calledWithContext]("foo");

      expect(() => {
        expect(fn).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches with a matcher", () => {
    const fn = vi.fn();
    fn.call("foo");

    if (isToHaveNth(calledWithContext)) {
      expect(fn)[calledWithContext](1, expect.any(String));

      expect(() => {
        expect(fn).not[calledWithContext](1, expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn)[calledWithContext](expect.any(String));

      expect(() => {
        expect(fn).not[calledWithContext](expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    }
  });

  it("includes the custom mock name in the error message", () => {
    const fn = vi.fn().mockName("named-mock");

    fn.call("foo");

    if (isToHaveNth(calledWithContext)) {
      expect(fn)[calledWithContext](1, "foo");

      expect(() => {
        expect(fn).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(fn)[calledWithContext]("foo");

      expect(() => {
        expect(fn).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
});
