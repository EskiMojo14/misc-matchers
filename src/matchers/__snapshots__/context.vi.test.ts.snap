// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lastCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`lastCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`lastCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`lastCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`lastCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`lastCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`lastCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>lastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`nthCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`nthCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`nthCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`nthCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`nthCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`nthCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`nthCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>nthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toBeCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.toBeCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>"foo"</>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toBeCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`toBeCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).toBeCalledWithContext(</><g>expected</><d>)</>

Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`toBeCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).toBeCalledWithContext(</><g>expected</><d>)</>

Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toBeCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).toBeCalledWithContext(</><g>expected</><d>)</>

Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toBeCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.toBeCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>"foo"</>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toBeCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.toBeCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>"foo"</>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`toHaveBeenCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`toHaveBeenCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>"foo"</>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.toHaveBeenCalledWithContext(</><g>expected</><d>)</>

Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenLastCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenLastCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`toHaveBeenLastCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>toHaveBeenLastCalledWithContext<d>(</><g>expected</><d>)</>

Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenNthCalledWithContext > includes the custom mock name in the error message 1`] = `
<d>expect(</><r>named-mock</><d>).</>not<d>.</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenNthCalledWithContext > works only on spies or vi.fn 1`] = `utils.printWithType is not a function`;

exports[`toHaveBeenNthCalledWithContext > works when not called 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>"foo"</>

Number of calls: <r>0</>
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that doesn't match 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>"foo"</>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that doesn't match with matchers 1`] = `
<d>expect(</><r>spy</><d>).</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: <g>Any<Number></>
Received: <r>"bar"</>
Number of calls: <r>1</>
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that matches 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>"foo"</>

Number of calls: <r>1</>
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that matches with a matcher 1`] = `
<d>expect(</><r>spy</><d>).</>not<d>.</>toHaveBeenNthCalledWithContext<d>(</>n<d>, </><g>expected</><d>)</>

n: 1
Expected: not <g>Any<String></>
Received:     <d>"foo"</>

Number of calls: <r>1</>
`;
